src/autopas/containers/CellBlock3D.h

```diff
@@ -106,127 +105,123 @@ class CellBlock3D : public CellBorderAndFlagManager {
+   * @param position the given position.
+   * @return true if the position is inside the halo region.
    */
   [[nodiscard]] bool checkInHalo(const std::array<double, 3> &position) const;
 
   /**
-   * deletes all particles in the halo cells of the managed cell block
+   * Deletes all particles in the halo cells of the managed cell block.
    */
   void clearHaloCells();
 
   /**
-   * Get the nearby halo cells.
-   * A list of halo cells is returned whose distance to position is at most
-   * allowedDistance. If position is inside a halo cell that cell is also
-   * returned. (1 norm is used, i.e. the distance is computed for each dimension
-   * separately, Manhattan distance)
-   * @param position cells close to this position are to be returned
-   * @param allowedDistance the maximal distance to the position
-   * @return a container of references to nearby halo cells
+   * Get the halo cells around a given point.
+   * Returns a list of halo cells that are at least partially within allowedDistance of the given position.
+   * If position is inside a halo cell that cell is also returned.
+   * @note The 1 norm is used, i.e. the distance is computed for each dimension separately, aka. Manhattan distance)
+   *
+   * @param position cells close to this position are to be returned.
+   * @param allowedDistance the maximal distance to the position.
+   * @return A vector of references to nearby halo cells.
    */
   std::vector<ParticleCell *> getNearbyHaloCells(const std::array<double, 3> &position, double allowedDistance) const {
     using namespace autopas::utils::ArrayMath::literals;
 
-    auto index3d = get3DIndexOfPosition(position);
+    const auto index3D = get3DIndexOfPosition(position);
+    const auto index1D = threeToOneD(index3D);
 
     std::vector<ParticleCell *> closeHaloCells;
-
-    auto isHaloCell = [&](const auto &index) {
-      for (size_t i = 0; i < 3; ++i) {
-        if (index[i] < _cellsPerInteractionLength or
-            index[i] >= _cellsPerDimensionWithHalo[i] - _cellsPerInteractionLength) {
-          return true;
-        }
-      }
-      return false;
-    };
-
-    // always add the cell the particle is currently in first, for that we test if it is a halo cell.
-    if (isHaloCell(index3d)) {
-      closeHaloCells.push_back(&getCell(index3d));
+    // x2 for left and right of the cell; +1 for the cell itself
+    const auto interestingCellsPerDim = utils::ArrayMath::ceilToInt((2. * allowedDistance / _cellLength)) + 1;
+    const auto interestingCellsBlockSize =
+        interestingCellsPerDim[0] * interestingCellsPerDim[1] * interestingCellsPerDim[2];
+    // This is an overestimation with the upper bound of possible number of cells in the vicinity.
+    // An overestimate is cheaper than many reallocations.
+    // If the size of the overallocation ever becomes a problem we can use vector::shrink_to_fit() before return.
+    closeHaloCells.reserve(interestingCellsBlockSize);
```

Comment:
I was wondering if we can reserve the size based on `lowIndex3D` and `highIndex3D` that we calculate anyway? Would that offer any benefit (because I am not sure)?

Also, this includes the current cell which we don't need to have, so we could instead have `closeHaloCells.reserve(interestingCellsBlockSize - 1);`? again, not sure if it makes much difference.

Reply:
Actually we can. I must have overlooked that when I implemented it, good catch!
-> 21fbfa391

The current cell might be a halo cell so it should be included in the estimate. Anyways it is better to slightly overestimate than to underestimate. Overestimation is basically free but underestimateion potentially costs a reallocation.